src/database/postgres_db.rs

```diff
@@ -0,0 +1,1253 @@
+        self.client
+            .borrow_mut()
+            .execute(&statement, &[&keychain, &(child as i32)])?;
+
+        Ok(())
+    }
+
+    fn delete_script_pubkey_by_script(&self, script: &[u8]) -> Result<(), Error> {
+        let statement = self
+            .client
+            .borrow_mut()
+            .prepare("DELETE FROM script_pubkeys WHERE script=$1")?;
+        self.client.borrow_mut().execute(&statement, &[&script])?;
+
+        Ok(())
+    }
+
+    fn delete_utxo_by_outpoint(&self, txid: &[u8], vout: u32) -> Result<(), Error> {
+        let statement = self
+            .client
+            .borrow_mut()
+            .prepare("DELETE FROM utxos WHERE txid=$1 AND vout=$2")?;
+        self.client
+            .borrow_mut()
+            .execute(&statement, &[&txid, &(vout as i32)])?;
+
+        Ok(())
+    }
+
+    fn delete_transaction_by_txid(&self, txid: &[u8]) -> Result<(), Error> {
+        let statement = self
+            .client
+            .borrow_mut()
+            .prepare("DELETE FROM transactions WHERE txid=$1")?;
+        self.client.borrow_mut().execute(&statement, &[&txid])?;
+        Ok(())
+    }
+
+    fn delete_transaction_details_by_txid(&self, txid: &[u8]) -> Result<(), Error> {
+        let statement = self
+            .client
+            .borrow_mut()
+            .prepare("DELETE FROM transaction_details WHERE txid=$1")?;
+        self.client.borrow_mut().execute(&statement, &[&txid])?;
+        Ok(())
+    }
+
+    fn delete_last_derivation_index_by_keychain(&self, keychain: String) -> Result<(), Error> {
+        let statement = self
+            .client
+            .borrow_mut()
+            .prepare("DELETE FROM last_derivation_indices WHERE keychain=$1")?;
+        self.client.borrow_mut().execute(&statement, &[&keychain])?;
+
+        Ok(())
+    }
+
+    fn delete_sync_time(&self) -> Result<(), Error> {
+        let statement = self
+            .client
+            .borrow_mut()
+            .prepare("DELETE FROM sync_time WHERE id = 0")?;
+        self.client.borrow_mut().execute(&statement, &[])?;
+        Ok(())
+    }
+}
+
+impl BatchOperations for PostgresDatabase {
+    fn set_script_pubkey(
+        &mut self,
+        script: &Script,
+        keychain: KeychainKind,
+        child: u32,
+    ) -> Result<(), Error> {
+        let keychain = serde_json::to_string(&keychain)?;
+        self.insert_script_pubkey(keychain, child, script.as_bytes())?;
+        Ok(())
+    }
+
+    fn set_utxo(&mut self, utxo: &LocalUtxo) -> Result<(), Error> {
+        self.insert_utxo(
+            utxo.txout.value,
+            serde_json::to_string(&utxo.keychain)?,
+            utxo.outpoint.vout,
+            &utxo.outpoint.txid,
+            utxo.txout.script_pubkey.as_bytes(),
+            utxo.is_spent,
+        )?;
+        Ok(())
+    }
+
+    fn set_raw_tx(&mut self, transaction: &Transaction) -> Result<(), Error> {
+        match self.select_transaction_by_txid(&transaction.txid())? {
```

Comment:
seems most of the `set_xxx` are using similiar pattern. the update is not atomic and there are no unique indexes on some tables.
we may try to synchronize multiple instances at the same time to see if there are any issues.

Reply:
Let me double check, I think that all operations are part of batch:

Reply:
Wallet for example uses `BatchDatabase`, I think the other traits are there to implement common functionality, but not to be used externally.

```
impl<D> Wallet<D>
where
    D: BatchDatabase,
{
...
```
```
/// Trait for a database that supports batch operations
///
/// This trait defines the methods to start and apply a batch of operations.
pub trait BatchDatabase: Database {
    /// Container for the operations
    type Batch: BatchOperations;

    /// Create a new batch container
    fn begin_batch(&self) -> Self::Batch;
    /// Consume and apply a batch of operations
    fn commit_batch(&mut self, batch: Self::Batch) -> Result<(), Error>;
}
```

But...

I don't think that higher level framework is using `batch.begin and batch.commit` :cold_sweat:

Reply:
Actually, It does, at least for syncing, but it is not on the same level, so it's hard to trace...

So it's like this:
Wallet(wallet_sync -> wallet_setup) -> ElectrumBlockchain/EsploraBlockchain(wallet_setup) -> script_sync.rs(satisfy->into_db_update) -> DB(begin_batch) ... ElectrumBlockchain/EsploraBlockchain ... -> DB(commit_batch)